### PR TITLE
feat: ftltest supports pubsub

### LINKED ISSUE #1598: ftltest support for PubSub

### LINKED ISSUE #1703: Add runtime safety checks when publishing to external topics
Added compile time checks here: https://github.com/TBD54566975/ftl/pull/1702

There are ways around the compile time checks, so we should also add runtime safety

### PR DESCRIPTION
**Pull request by matt2e**
closes #1598

Adds:
- By default, disables all subscribers during unit tests
- Subscribers can be manually added per unit test
   - Subscribers can be actual sinks, or ad-hoc ones defined in the test (just define a `func` in the argument)
- Check what events were published to a topic
- Check what events were consumed by a subscription, and whether an error occured
- Allow the unit test to wait for all events to be consumed
- Unit tests can publish to external module's topics
    - And conversely, added a runtime check to disallow publishing to external modules when not in a unit test. fixes #1703

### CODE REVIEW

File: go-runtime/ftl/pubsub.go [hunk 1]
```diff
@@ -3,32 +3,38 @@ package ftl
...
 }
 
 // Publish publishes an event to a topic
 func (t TopicHandle[E]) Publish(ctx context.Context, event E) error {
-	return internal.FromContext(ctx).PublishEvent(ctx, t.name, event)
+	return internal.FromContext(ctx).PublishEvent(ctx, t.Ref, event)
+}
+
+type SubscriptionHandle[E any] struct {
+	Topic *schema.Ref
```
Comment (matt2e) [inline go-runtime/ftl/pubsub.go:31]: TopicHandle and SubscriptionHandle's fields are accessed by ftltest, so have been made public

File: go-runtime/ftl/ftltest/fake.go [hunk 2]
```diff
@@ -4,33 +4,97 @@ import (
...
-func newFakeFTL() *fakeFTL {
-	return &fakeFTL{
+func newFakeFTL(ctx context.Context) *fakeFTL {
+	fake := &fakeFTL{
 		fsm:           newFakeFSMManager(),
 		mockMaps:      map[uintptr]mapImpl{},
 		allowMapCalls: false,
 		configValues:  map[string][]byte{},
 		secretValues:  map[string][]byte{},
+		globalTopic:   pubsub.New[pubSubEvent](),
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:89]: Let's pull all the pubsub code out into a separate struct and file, and forward to it from fakeFTL.

File: go-runtime/ftl/ftltest/fake.go [hunk 3]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+func (f *fakeFTL) waitForSubscriptionsToComplete(ctx context.Context) {
+	logger := log.FromContext(ctx).Scope("pubsub")
+	startTime := time.Now()
+	nextLogTime := startTime.Add(2 * time.Second)
+
+	// Make sure all published events make it into our pubsub state
+	f.publishWaitGroup.Wait()
+
+	for {
+		if func() bool {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:362]: This is weird. Let's pull the body out into its own method.

File: go-runtime/ftl/ftltest/fake.go [hunk 4]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
 }
+
+func (f *fakeFTL) PublishEvent(ctx context.Context, topic *schema.Ref, event any) error {
+	f.publishWaitGroup.Add(1)
+	f.globalTopic.PublishSync(publishEvent{topic: topic, content: event})
+	return nil
+}
+
+// addSubscriber adds a subscriber to the fake FTL instance. Each subscriber included in the test must be manually added
+func addSubscriber[E any](f *fakeFTL, sub ftl.SubscriptionHandle[E], sink ftl.Sink[E]) {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:192]: This should just be a method.

File: go-runtime/ftl/ftltest/fake.go [hunk 5]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+	f.subscribers[sub.Name] = append(subscribers, func(ctx context.Context, event any) error {
+		if event, ok := event.(E); ok {
+			return sink(ctx, event)
+		}
+		return fmt.Errorf("unexpected event type %T for subscription %s", event, sub.Name)
+	})
+}
+
+// eventsForTopic returns all events published to a topic
+func eventsForTopic[E any](ctx context.Context, f *fakeFTL, topic ftl.TopicHandle[E]) []E {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:217]: Same here, should be a method.

File: go-runtime/ftl/ftltest/fake.go [hunk 6]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+			events = append(events, event)
+		} else {
+			logger.Warnf("unexpected event type %T for topic %s", e, topic.Ref)
+		}
+	}
+	return events
+}
+
+// resultsForSubscription returns all consumed events and whether an error was returned
+func resultsForSubscription[E any](ctx context.Context, f *fakeFTL, handle ftl.SubscriptionHandle[E]) []SubscriptionResult[E] {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:241]: And here.

File: go-runtime/ftl/ftltest/fake.go [hunk 7]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+			topic:  sub.Topic,
+			errors: map[int]error{},
+		}
+	}
+
+	subscribers, ok := f.subscribers[sub.Name]
+	if !ok {
+		subscribers = []subscriber{}
+	}
+	f.subscribers[sub.Name] = append(subscribers, func(ctx context.Context, event any) error {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:208]:
```suggestion
	f.subscribers[sub.Name] = append(f.subscribers[sub.Name], func(ctx context.Context, event any) error {
```

^^ this is equivalent

File: go-runtime/ftl/ftltest/fake.go [hunk 8]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+	return events
+}
+
+// resultsForSubscription returns all consumed events and whether an error was returned
+func resultsForSubscription[E any](ctx context.Context, f *fakeFTL, handle ftl.SubscriptionHandle[E]) []SubscriptionResult[E] {
+	f.pubSubLock.Lock()
+	defer f.pubSubLock.Unlock()
+
+	logger := log.FromContext(ctx).Scope("pubsub")
+	results := []SubscriptionResult[E]{}
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:246]:
```suggestion
	var results []SubscriptionResult[E]
```

tip: avoids an allocation

File: go-runtime/ftl/ftltest/fake.go [hunk 9]
```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
...
+				})
+			}
+			if len(remaining) == 0 {
+				// not waiting on any more subscriptions
+				return true
+			}
+			if time.Now().After(nextLogTime) {
+				// print out what we are waiting on
+				nextLogTime = time.Now().Add(2 * time.Second)
+				logger.Infof("waiting on subscriptions to complete after %ds:\n%s", int(time.Until(startTime).Seconds()*-1), strings.Join(slices.Map(remaining, func(r remainingState) string {
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/fake.go:401]: This should be `Debugf` or `Tracef` no?

File: go-runtime/ftl/ftltest/ftltest.go [hunk 10]
```diff
@@ -371,3 +371,61 @@ func getDSNFromSecret(ftl internal.FTL, module, name string) (string, error) {
...
+
+// EventsForTopic returns all published events for a topic
+func EventsForTopic[E any](ctx context.Context, topic ftl.TopicHandle[E]) []E {
+	fftl := internal.FromContext(ctx).(*fakeFTL) //nolint:forcetypeassert
+	return eventsForTopic(ctx, fftl, ftl.TopicHandle[E](topic))
+}
+
+type SubscriptionResult[E any] struct {
+	Event E
+	Error error
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/ftltest.go:404]: Can this be nil? If so it should be made into an `Option[error]` instead

File: go-runtime/ftl/ftltest/testdata/go/pubsub/go.mod [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	golang.org/x/exp v0.0.0-20240506185415-9bf2ced13842 // indirect
+	golang.org/x/mod v0.18.0 // indirect
+	golang.org/x/net v0.26.0 // indirect
+	golang.org/x/sync v0.7.0 // indirect
+	golang.org/x/sys v0.21.0 // indirect
+	golang.org/x/text v0.16.0 // indirect
+	google.golang.org/protobuf v1.34.1 // indirect
+)
+
+replace github.com/TBD54566975/ftl => ./../../../../../..
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/testdata/go/pubsub/go.mod:64]: That's a lotta dots

**Review by alecthomas — APPROVED** [review]

**Pull request merged**